### ISSUE TITLE
SDDM does not start and not log anything when using wayland

### ISSUE THREAD

**Issue by PucklaJ** [sddm/sddm#1807]
I was using SDDM under wayland (as configured in the Arch Wiki article) completely fine, but suddenly after a system update it stopped working. Now I just get a blinking cursor. I tried to look into the logs, but it does not generate any logs at all. I am starting SDDM through dinit and setup a logfile through the service scripts. It does create a log file but does not write a single byte into it. If I switch back to x11 it works completely fine.

Here is some system information:
OS: Artix Linux
Kernel: 6.1.54-1-lts
CPU: Intel Xeon E5-2620 v4 (16) @ 3.000GHz
GPU: AMD ATI Radeon R9 285/380

SDDM: 0.20.0-2
weston: 12.0.2-1

`/etc/sddm.conf.d/10-wayland.conf`:
```
[General]
DisplayServer=wayland
```

`/etc/sddm.conf.d/20-autologin.conf`:
```
#[Autologin]
#Relogin=false
#User=username
#Session=hyprland-dbus
```

`/etc/sddm.conf.d/30-theme.conf`:
```
[Theme]
Current=Apple-sddm-v1
CursorTheme=Dracula-cursors
CursorSize=24
```

`/etc/dinit.d/sddm`:
```
type            = process
command         = /usr/bin/sddm
smooth-recovery = true
depends-on      = logind
depends-on      = sddm-pre
after           = loginready
waits-for.d     = live.d
logfile         = /var/log/dinit/sddm.log
```

`/etc/dinit.d/sddm-pre`:
```
type    = scripted
command = /usr/bin/install -dm0755 -o sddm -g sddm /var/lib/sddm
after   = root-rw
```

**Comment by PucklaJ** [comment]
I retrieved a log after starting it manually:
```
[06:45:22.093] (II) DAEMON: Initializing...
[06:45:22.098] (II) DAEMON: Starting...
[06:45:22.098] (II) DAEMON: Logind interface found
[06:45:22.099] (II) DAEMON: Adding new display...
[06:45:22.099] (II) DAEMON: Loaded empty theme configuration
[06:45:22.100] (II) DAEMON: Using VT 1
[06:45:22.100] (II) DAEMON: Display server started.
[06:45:22.100] (II) DAEMON: Socket server starting...
[06:45:22.101] (II) DAEMON: Socket server started.
[06:45:22.101] (II) DAEMON: Loading theme configuration from "/usr/share/sddm/themes/Apple-sddm-v1/theme.conf"
[06:45:22.101] (II) DAEMON: Greeter starting...
[06:45:22.117] (II) HELPER: [PAM] Starting...
[06:45:22.117] (II) HELPER: [PAM] Authenticating...
[06:45:22.117] (II) HELPER: [PAM] returning.
[06:45:22.120] (II) DAEMON: Greeter session started successfully
[06:45:22.121] (II) HELPER: [PAM] Closing session
[06:45:22.121] (II) HELPER: [PAM] Ended.
[06:45:22.122] (WW) DAEMON: Auth: sddm-helper exited with 5
[06:45:22.122] (II) DAEMON: Greeter stopped. SDDM::Auth::HELPER_TTY_ERROR
[06:45:22.122] (II) DAEMON: Jumping to VT 1
[06:45:22.122] (II) DAEMON: VT mode didn't need to be fixed
[06:45:22.122] (II) DAEMON: Socket server stopping...
[06:45:22.122] (II) DAEMON: Socket server stopped.
[06:45:22.122] (II) DAEMON: Removing display SDDM::Display(0x55e0496b1f60) ...
[06:45:22.123] (II) DAEMON: Adding new display...
[06:45:22.123] (II) DAEMON: Loaded empty theme configuration
[06:45:22.137] (II) DAEMON: Using VT 1
[06:45:22.137] (II) DAEMON: Display server started.
[06:45:22.137] (II) DAEMON: Socket server starting...
[06:45:22.137] (II) DAEMON: Socket server started.
[06:45:22.137] (II) DAEMON: Loading theme configuration from "/usr/share/sddm/themes/Apple-sddm-v1/theme.conf"
[06:45:22.137] (II) DAEMON: Greeter starting...
[06:45:22.153] (II) HELPER: [PAM] Starting...
[06:45:22.153] (II) HELPER: [PAM] Authenticating...
[06:45:22.153] (II) HELPER: [PAM] returning.
[06:45:22.157] (II) DAEMON: Greeter session started successfully
[06:45:22.157] (II) HELPER: [PAM] Closing session
[06:45:22.157] (II) HELPER: [PAM] Ended.
[06:45:22.158] (WW) DAEMON: Auth: sddm-helper exited with 5
[06:45:22.158] (II) DAEMON: Greeter stopped. SDDM::Auth::HELPER_TTY_ERROR
[06:45:22.158] (II) DAEMON: Jumping to VT 1
[06:45:22.201] (II) DAEMON: VT mode didn't need to be fixed
[06:45:22.202] (II) DAEMON: Socket server stopping...
[06:45:22.202] (II) DAEMON: Socket server stopped.
[06:45:22.202] (II) DAEMON: Removing display SDDM::Display(0x55e0496b2fe0) ...
[06:45:22.202] (II) DAEMON: Adding new display...
[06:45:22.202] (II) DAEMON: Loaded empty theme configuration
[06:45:22.204] (II) DAEMON: Using VT 1
[06:45:22.204] (II) DAEMON: Display server started.
[06:45:22.204] (II) DAEMON: Socket server starting...
[06:45:22.204] (II) DAEMON: Socket server started.
[06:45:22.204] (II) DAEMON: Loading theme configuration from "/usr/share/sddm/themes/Apple-sddm-v1/theme.conf"
[06:45:22.205] (II) DAEMON: Greeter starting...
[06:45:22.221] (II) HELPER: [PAM] Starting...
[06:45:22.221] (II) HELPER: [PAM] Authenticating...
[06:45:22.221] (II) HELPER: [PAM] returning.
[06:45:22.225] (II) DAEMON: Greeter session started successfully
[06:45:22.225] (II) HELPER: [PAM] Closing session
[06:45:22.225] (II) HELPER: [PAM] Ended.
[06:45:22.226] (WW) DAEMON: Auth: sddm-helper exited with 5
[06:45:22.226] (II) DAEMON: Greeter stopped. SDDM::Auth::HELPER_TTY_ERROR
[06:45:22.226] (II) DAEMON: Jumping to VT 1
[06:45:22.285] (II) DAEMON: VT mode didn't need to be fixed
[06:45:22.285] (II) DAEMON: Socket server stopping...
[06:45:22.285] (II) DAEMON: Socket server stopped.
[06:45:22.285] (II) DAEMON: Removing display SDDM::Display(0x55e0496b1f60) ...
[06:45:22.285] (II) DAEMON: Adding new display...
[06:45:22.285] (II) DAEMON: Loaded empty theme configuration
[06:45:22.287] (II) DAEMON: Using VT 1
[06:45:22.287] (II) DAEMON: Display server started.
[06:45:22.287] (II) DAEMON: Socket server starting...
[06:45:22.287] (II) DAEMON: Socket server started.
[06:45:22.287] (II) DAEMON: Loading theme configuration from "/usr/share/sddm/themes/Apple-sddm-v1/theme.conf"
[06:45:22.288] (II) DAEMON: Greeter starting...
[06:45:22.305] (II) HELPER: [PAM] Starting...
[06:45:22.305] (II) HELPER: [PAM] Authenticating...
[06:45:22.305] (II) HELPER: [PAM] returning.
[06:45:22.309] (II) DAEMON: Greeter session started successfully
[06:45:22.309] (II) HELPER: [PAM] Closing session
[06:45:22.309] (II) HELPER: [PAM] Ended.
[06:45:22.310] (WW) DAEMON: Auth: sddm-helper exited with 5
[06:45:22.310] (II) DAEMON: Greeter stopped. SDDM::Auth::HELPER_TTY_ERROR
[06:45:22.310] (II) DAEMON: Jumping to VT 1
[06:45:22.385] (II) DAEMON: VT mode didn't need to be fixed
[06:45:22.385] (II) DAEMON: Socket server stopping...
[06:45:22.385] (II) DAEMON: Socket server stopped.
[06:45:22.385] (II) DAEMON: Removing display SDDM::Display(0x55e0496b5830) ...
[06:45:22.385] (II) DAEMON: Adding new display...
[06:45:22.385] (II) DAEMON: Loaded empty theme configuration
[06:45:22.387] (II) DAEMON: Using VT 1
[06:45:22.387] (II) DAEMON: Display server started.
[06:45:22.387] (II) DAEMON: Socket server starting...
[06:45:22.387] (II) DAEMON: Socket server started.
[06:45:22.388] (II) DAEMON: Loading theme configuration from "/usr/share/sddm/themes/Apple-sddm-v1/theme.conf"
[06:45:22.388] (II) DAEMON: Greeter starting...
[06:45:22.408] (II) HELPER: [PAM] Starting...
[06:45:22.408] (II) HELPER: [PAM] Authenticating...
[06:45:22.408] (II) HELPER: [PAM] returning.
[06:45:22.412] (II) DAEMON: Greeter session started successfully
[06:45:22.412] (II) HELPER: [PAM] Closing session
[06:45:22.413] (II) HELPER: [PAM] Ended.
[06:45:22.414] (WW) DAEMON: Auth: sddm-helper exited with 5
[06:45:22.414] (II) DAEMON: Greeter stopped. SDDM::Auth::HELPER_TTY_ERROR
[06:45:22.414] (II) DAEMON: Jumping to VT 1
[06:45:22.501] (II) DAEMON: VT mode didn't need to be fixed
[06:45:22.501] (II) DAEMON: Socket server stopping...
[06:45:22.501] (II) DAEMON: Socket server stopped.
[06:45:22.502] (II) DAEMON: Removing display SDDM::Display(0x55e0496b1b20) ...
[06:45:22.502] (II) DAEMON: Adding new display...
[06:45:22.502] (II) DAEMON: Loaded empty theme configuration
[06:45:22.504] (II) DAEMON: Using VT 1
[06:45:22.504] (II) DAEMON: Display server started.
[06:45:22.504] (II) DAEMON: Socket server starting...
[06:45:22.505] (II) DAEMON: Socket server started.
[06:45:22.505] (II) DAEMON: Loading theme configuration from "/usr/share/sddm/themes/Apple-sddm-v1/theme.conf"
[06:45:22.505] (II) DAEMON: Greeter starting...
[06:45:22.524] (II) HELPER: [PAM] Starting...
[06:45:22.524] (II) HELPER: [PAM] Authenticating...
[06:45:22.524] (II) HELPER: [PAM] returning.
[06:45:22.528] (II) DAEMON: Greeter session started successfully
[06:45:22.528] (II) HELPER: [PAM] Closing session
[06:45:22.529] (II) HELPER: [PAM] Ended.
[06:45:22.530] (WW) DAEMON: Auth: sddm-helper exited with 5
[06:45:22.530] (II) DAEMON: Greeter stopped. SDDM::Auth::HELPER_TTY_ERROR
[06:45:22.530] (II) DAEMON: Jumping to VT 1
[06:45:22.618] (II) DAEMON: VT mode didn't need to be fixed
[06:45:22.618] (II) DAEMON: Socket server stopping...
[06:45:22.618] (II) DAEMON: Socket server stopped.
[06:45:22.618] (II) DAEMON: Removing display SDDM::Display(0x55e0496b8270) ...
[06:45:22.619] (II) DAEMON: Adding new display...
[06:45:22.619] (II) DAEMON: Loaded empty theme configuration
[06:45:22.621] (II) DAEMON: Using VT 1
[06:45:22.621] (II) DAEMON: Display server started.
[06:45:22.621] (II) DAEMON: Socket server starting...
[06:45:22.621] (II) DAEMON: Socket server started.
[06:45:22.622] (II) DAEMON: Loading theme configuration from "/usr/share/sddm/themes/Apple-sddm-v1/theme.conf"
[06:45:22.622] (II) DAEMON: Greeter starting...
[06:45:22.624] (II) DAEMON: Socket server stopping...
[06:45:22.624] (II) DAEMON: Socket server stopped.
[06:45:22.624] (WW) DAEMON: Auth: sddm-helper (--socket /tmp/sddm-auth-1b2f0d2f-eb1a-47fd-90f4-005188ab00a6 --id 14 --start /usr/bin/sddm-greeter --socket /tmp/sddm--zMGZZa --theme /usr/share/sddm/themes/Apple-sddm-v1 -style System --user sddm --display-server weston --shell=fullscreen-shell.so --greeter) crashed (exit code 1)
[06:45:22.624] (WW) DAEMON: Auth: sddm-helper exited with 15
```

**Comment by BeyondMagic** [comment]
Same error occurring on also Artix Linux, but I'm running _runit_. The same logs, as shown by Jonas Pucher, are utter useless and impossible to even have a clue where to look at this source code to find and fix it.

Developers, please at the very least, give us a hint so that we can make SDDM better...

**Comment by JamesMcMahon** [comment]
Is this connected to https://github.com/sddm/sddm/issues/1592? I was hitting a similar issue and was fixed when I switched away from fish as my default shell.

**Comment by Izooc** [comment]
Same thing here on void linux. Seems to be a runit system problem maybe? I tried with regular Weston and kwin and I still have a blinking cursor. Not using fish as shell though, only bash...

**Comment by BeyondMagic** [comment]
> Seems to be a runit system problem maybe?

Not only runit fails as shown by the author of the post, although you may be on something since only non-systemd users have reported so far.

___

Looking at the code ~again~, it seems that the auth system is failling somehow even though PAM confirm being correctly authenticated in the only logs we can retrieve.

**Comment by BeyondMagic** [comment]
> Is this connected to https://github.com/sddm/sddm/issues/1592?

I don't believe so, `wayland-session` should only run after the user has logged in, the issue happens while trying to run the greeter.

---

Probably related to #1844.

**Comment by NyaomiDEV** [comment]
> Probably related to #1844.

It is the same issue.

**Comment by agowa** [comment]
Duplicates #1844 

I managed to debug this issue and find a workaround as well as what would have to be done for a permanent fix. I've posted it within the other ticket just now.

> I managed to nail down the root cause of this issue.
> 
> It's a wird edge case interaction. Thanks goes to [@Vogtinator](https://github.com/Vogtinator) who provided enough information to get starting debugging with especially this comment here [#1896 (comment)](https://github.com/sddm/sddm/pull/1896#issuecomment-2004263424)
> 
> > Apparently `getty@.service` has `Type=idle` which means it is marked as active before anything actually happened. I don't see a way to fix this.
> 
> (The entire journey of trial and error so far: https://chaos.social/@agowa338/114853300563852266 )
> 
> Tl;Dr: sddm should use the TTY capabilities of systemd but currently doesn't and trying to do it the same way as e.g. agetty results in sddm blocking the terminal itself and sddm-helper failing literally the same way.
> 
> the issue could have been avoided with e.g. this systemd unit (first one is what was packaged with sddm, 2nd is what I created to resolve this issue here. However there is some special handling that would be desirable for us that currenctly cant be used (`StandardInput=tty`, and `StandardOutput=tty` https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#StandardInput=).
> 
> > **StandardInput=**: (...)
> > If null is selected, standard input will be connected to /dev/null, i.e. all read attempts by the process will result in immediate EOF.
> > If tty is selected, standard input is connected to a TTY (as configured by TTYPath=, see below) and the executed process becomes the controlling process of the terminal. If the terminal is already being controlled by another process, the executed process waits until the current controlling process releases the terminal.
> > tty-force is similar to tty, but the executed process is forcefully and immediately made the controlling process of the terminal, potentially removing previous controlling processes from the terminal.
> > tty-fail is similar to tty, but if the terminal already has a controlling process start-up of the executed process fails.
> > (...)
> > This setting defaults to null, unless StandardInputText=/StandardInputData= are set, in which case it defaults to data.
> 
> Workaround that mostly works but requires knowing each and every other unit by name that is potentially using /dev/tty1.
> 
> ```
> [root@PC-001 ~]# systemctl cat sddm.service
> # /usr/lib/systemd/system/sddm.service
> [Unit]
> Description=Simple Desktop Display Manager
> Documentation=man:sddm(1) man:sddm.conf(5)
> Conflicts=getty@tty1.service
> After=systemd-user-sessions.service getty@tty1.service plymouth-quit.service systemd-logind.service
> PartOf=graphical.target
> StartLimitIntervalSec=30
> StartLimitBurst=2
> 
> [Service]
> ExecStart=/usr/bin/sddm
> Restart=always
> 
> [Install]
> Alias=display-manager.service
> 
> 
> # /etc/systemd/system/sddm.service.d/override.conf
> [Unit]
> Conflicts=console-getty.service
> After=console-getty.service
> ConditionPathExists=/dev/tty1
> 
> [Service]
> RestartSec=0
> StandardInput=null
> StandardOutput=journal
> TTYPath=/dev/tty1
> TTYReset=yes
> TTYVHangup=yes
> IgnoreSIGPIPE=no
> SendSIGHUP=yes
> ImportCredential=tty.tty1.agetty.*:agetty.
> ImportCredential=tty.tty1.login.*:login.
> ImportCredential=agetty.*
> ImportCredential=login.*
> ImportCredential=shell.*
> ```
> 
> What is needed to remediate this issue entirely is:
> 
> * sddm needs to be able to accept a tty device on stdin, stdout, and stderr (at a minimum at stdin for the special locking mechanism of systemd to take effect).
> * sddm must handle receiving a SIGHUP which means it lost access to the tty to some other service unit.
> * afterwards this sddm.service unit options should be used (doesn't need to know all other units by name that may be using /dev/tty1):
> 
> ```
> # /usr/lib/systemd/system/sddm.service
> [Unit]
> Description=Simple Desktop Display Manager
> Documentation=man:sddm(1) man:sddm.conf(5)
> After=systemd-user-sessions.service plymouth-quit.service systemd-logind.service getty-pre.target
> Before=getty.target
> PartOf=graphical.target
> StartLimitIntervalSec=30
> StartLimitBurst=2
> 
> ConditionPathExists=/dev/tty1
> 
> [Service]
> ExecStart=/usr/bin/sddm
> Restart=always
> RestartSec=0
> StandardInput=tty
> StandardOutput=tty
> UtmpIdentifier=tty1
> TTYPath=/dev/tty1
> TTYReset=yes
> TTYVHangup=yes
> IgnoreSIGPIPE=no
> SendSIGHUP=yes
> ImportCredential=tty.tty1.agetty.*:agetty.
> ImportCredential=tty.tty1.login.*:login.
> ImportCredential=agetty.*
> ImportCredential=login.*
> ImportCredential=shell.*
> 
> [Install]
> WantedBy=getty.target
> Alias=display-manager.service
> ```
> 
> While testing I didn't have xorg-server installed and had my `/etc/sddm.conf.d/settings.conf` set to:
> 
> ```
> [General]
> DisplayServer=wayland
> GreeterEnvironment=QT_WAYLAND_SHELL_INTEGRATION=layer-shell
> 
> [Wayland]
> CompositorCommand=kwin_wayland --drm --no-lockscreen --no-global-shortcuts --locale1
> 
> [Theme]
> EnableAvatars=false
> ```

**Comment by thePrivacyFanatic** [comment]
> Apparently getty@.service has Type=idle

won't really help considering this error literally appears only without systemd
from testing it seems the wayland mode (with kwin) works just fine but this means nothing for the problem at hand

but you are right that letting it take a chosen tty instead of setting it at compile time could help
if memory serves older versions would just loop over all ttys to look for a free one but now it asks logind (which is replaced by a stub version called elogind in non systemd distros that only provides minimum functionality)
this means that maybe if the code to look for it normally still exists this just needs a method to know if logind or elogind is being used and act accordingly.
